instrument/call.go

```diff
@@ -65,3 +65,18 @@ func (c *call) Exec(f ExecFn) error {
 
 	return nil
 }
+
```

Comment:
could you re-implement `Exec` to use `ExecWithFilter` internally

Reply:
We absolutely could and I considered this. However, `Exec` is most likely the common case for most people, and if we implement it with `ExecWithFilter` we're now going to be incurring two extra function calls (one to `ExectWithFilter` itself, and then one to a default filter that always returns false). I'm not a performance expert, but I was worried about how this would affect performance for the common case.

Will Go just inline the function calls anyway? Want me to do any benchmarking? Or am I just over thinking things here and we should just implement `Exec` by calling `ExecWithFilter`?

Reply:
I wrote a quick benchmark and have some not too surprising results. I implemented an `Exec2` function that just calls `ExecWithFilter` as you suggested. I then wrote the following benchmarks (where `Exec` is just the current `Exec` left as is:

```go

func BenchmarkExec(b *testing.B) {
	s := tally.NewTestScope("", nil)
	instrumented := NewCall(s, "test_call")
	for n := 0; n < b.N; n++ {
		instrumented.Exec(func() error {
			if rand.Intn(1000) == 999 {
				return errors.New("expected")
			}
			return nil
		})
	}
}

func BenchmarkExec2(b *testing.B) {
	s := tally.NewTestScope("", nil)
	instrumented := NewCall(s, "test_call")
	for n := 0; n < b.N; n++ {
		instrumented.Exec2(func() error {
			if rand.Intn(1000) == 999 {
				return errors.New("expected")
			}
			return nil
		})
	}
}
```
Basically, call the instrumented function a bunch of times, and every so often have the function return an error (1 in a 1000 times to be exact).

Here are the results from three runs of the benchmark:
```
kurtis-C02T515QHF1T:instrument kurtisnusbaum$ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/uber-go/tally/instrument
BenchmarkExec-4          1000000              1604 ns/op
BenchmarkExec2-4         1000000              1653 ns/op
PASS
ok      github.com/uber-go/tally/instrument     3.323s
kurtis-C02T515QHF1T:instrument kurtisnusbaum$ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/uber-go/tally/instrument
BenchmarkExec-4          1000000              1587 ns/op
BenchmarkExec2-4         1000000              1613 ns/op
PASS
ok      github.com/uber-go/tally/instrument     3.265s
kurtis-C02T515QHF1T:instrument kurtisnusbaum$ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/uber-go/tally/instrument
BenchmarkExec-4          1000000              1612 ns/op
BenchmarkExec2-4         1000000              1678 ns/op
PASS
ok      github.com/uber-go/tally/instrument     3.355s
```
As you can see, `Exec2` is just every so slightly slow in a pretty consistent fashion. It looks like on average `Exec` as is runs about 50 ns faster. I guess the question now is, do we value the code deduplication more than we value the 50 ns. Let me know what you think and I'll make the change.

Reply:
tl;dr - yeah I don't think the 50ns matter for this case. We're not anywhere nearly that frequency for metric sampling. Go ahead and rewrite into the new version.

Longer story –

I did a benchmark to see function execution overhead and it was pretty much nothing.
```
func call(err error) error {
    return err
}
 
func BenchmarkInlining(b *testing.B) {
    err := fmt.Errorf("random err")
    b.ReportAllocs()
    for i := 0; i < b.N; i++ {
        err = call(err)
    }
}
```

Results:
```
    goos: darwin
    goarch: amd64
    BenchmarkInlining-4     2000000000           0.33 ns/op        0 B/op          0 allocs/op
    PASS
 ```

There are a couple of things different about your benchmark than the final version would look like:
- allocate the passthru lambda on the stack, rather than inline 
- your lambda is doing things beyond just passthru

Would be interesting to see the results of your benchmark with these changes